src/pyscipopt/scip.pyx

```diff
@@ -4140,11 +4144,9 @@ cdef class Model:
         """
         if sol == None:
             sol = Solution.create(self._scip, NULL)
-        if isinstance(expr, Variable):
-            var = <Variable> expr
-            return SCIPgetSolVal(self._scip, sol.sol, var.scip_var)
-        else:
-            return expr._evaluate(sol)
+        if self._scip != (<Solution>sol).scip:
```

Comment:
The idea behind the change in `getSolVal` is to make `Solution.__getitem__` the only one responsible for retrieving values for any `Expr`. However,  it appears that sometimes `sol.scip` is out of sync with `self._scip` and causes a segmentation fault error. I've observed this in https://github.com/SCIP-Interfaces/PySCIPOpt/blob/3fea23e9c993cdfea3e4f13f52463e29d5fbf735/tests/test_alldiff.py#L184 and https://github.com/SCIP-Interfaces/PySCIPOpt/blob/3fea23e9c993cdfea3e4f13f52463e29d5fbf735/tests/test_tsp.py#L23.
I suspect that the reason has something to do with how references to model are kept in the different plugins https://github.com/SCIP-Interfaces/PySCIPOpt/blob/3fea23e9c993cdfea3e4f13f52463e29d5fbf735/src/pyscipopt/scip.pyx#L3384 but I don't know how to fix it. @mattmilten @fserra, do you have any suggestions?

This makes me think that maybe the `Model` class should be the one in charge of evaluating `Expr` instead of `Solution`. In this case, `Solution` would need some back reference to the model. Another alternative would be to repeat the logic both in `Solution.__getitem__` and `Model.getSolVal` but I would prefer to avoid that if possible.

Reply:
Hmm, I don't see the issue here. All tests are running fine, right?

Also, please excuse the delay in replying!

Reply:
mmm I think this shouldn't be happening. It is a bit inefficient to create a python object to wrap the solution every time one accesses a value. e.g. from test_tsp
```
        for (i, j) in x:
            if self.model.getSolVal(solution, x[i, j]) > EPS:
```
For that particular code, I think the problem is that in `conshdlr.pxi` we do

```
cdef SCIP_RETCODE PyConsCheck (SCIP* scip, SCIP_CONSHDLR* conshdlr, SCIP_CONS** conss, int nconss, SCIP_SOL* sol, SCIP_Bool checkintegrality,
                               SCIP_Bool checklprows, SCIP_Bool printreason, SCIP_Bool completely, SCIP_RESULT* result):
    PyConshdlr = getPyConshdlr(conshdlr)
    cdef constraints = []
    for i in range(nconss):
        constraints.append(getPyCons(conss[i]))
    solution = Solution()
    solution.sol = sol
    result_dict = PyConshdlr.conscheck(constraints, solution, checkintegrality, checklprows, printreason, completely)
    result[0] = result_dict.get("result", <SCIP_RESULT>result[0])
    return SCIP_OKAY
```
So when the solution is created, we do not set its `scip` field. Maybe that would be enough to remove the second if.

Regarding the first if, i.e.,
```
        if sol == None:
            sol = Solution.create(self._scip, NULL)
```
We could just call directly `SCIPvarGetSol` instead of creating a solution object. (anyway SCIP should end up calling `SCIPvarGetSol` internally, unless I am missing something)

Reply:
@fserra, you are absolutely right, the problem was that the `scip` field was not being set. I've seen a few other places with the same issue and I've updated them to use `Solution.create(scip, sol)`. Regarding `SCIPvarGetSol` I'm not sure what you mean, as far as I can tell the function returns a real number https://scipopt.org/doc-7.0.2/html/group__PublicVariableMethods.php#ga110d0d9fad1f59b009126a84add117a8.

Reply:
I can't find the code anymore ;) but let me comment in another change what I mean